formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/TestFunctions.kt

```diff
@@ -24,9 +25,74 @@ inline fun <reified T : Throwable> assertFailsWithMessage(
     assertionMessage: String? = null,
     block: () -> Unit
 ) {
-    val exception = assertFailsWith(T::class, assertionMessage, block)
-    assertTrue(
-        exception.message!!.contains(message),
-        "expected:<$message> but was:<${exception.message}>"
+    assertFailsWith<T>(
+        assertionMessage,
+        {
+            assertFailsWith(message)
+        },
+        block,
     )
+}
+
+@DslMarker
+annotation class ExceptionCheckDsl
+
+@ExceptionCheckDsl
+interface ExceptionCheckScope<T> {
+    fun assertFailsWith(vararg message: String)
+    fun <R : Throwable> assertCausedBy(byType: KClass<R>, assertion: ExceptionCheckScope<R>.() -> Unit)
+}
+
+@ExceptionCheckDsl
+inline fun <reified R : Throwable> ExceptionCheckScope<*>.assertCausedBy(noinline assertion: ExceptionCheckScope<R>.() -> Unit) {
+    assertCausedBy(R::class, assertion)
+}
+
+inline fun <reified T : Throwable> assertFailsWith(
+    assertionMessage: String? = null,
+    assertion: ExceptionCheckScope<T>.() -> Unit = {},
+    block: () -> Unit
+) {
+    val exception = assertFailsWith(T::class, assertionMessage, block = block)
+    val scope = buildExceptionCheckScope(exception)
+    scope.assertion()
+}
+
+fun <T : Throwable> buildExceptionCheckScope(exception: T, depth: Int = 0): ExceptionCheckScope<T> = object : ExceptionCheckScope<T> {
+    override fun assertFailsWith(vararg message: String) {
+        val exceptionStackSize = exception.exceptionStackSize
+        assertTrue(message.size <= exceptionStackSize, "Expected exception to be assembled by ${message.size} throwable, but it has $exceptionStackSize, is $exception")
+        var index = 0
+        var currentException: Throwable? = exception
+        while (index < message.size) {
+            val currentMessage = message[index]
+            assertNotNull(currentException, "Expected exception to have a cause with message $currentMessage, but it was null")
+            require(currentException != null)
+            assertEquals(currentMessage, currentException.message, "Expected exception to have a cause with message <$currentMessage>, but it was <${currentException.message}> at cause stack ${index + depth}")
```

Comment:
After diving into `assertEquals` I found that it would append the `Expected <$expected>, actual <$actual>.` string after message passing in. So I would just simplify the input message.

It works in my IntelliJ IDEA 2024.2.1 RC and you can comment out the `Ignore` annotation in the new `TestFunctionTest` to see if it also works in other environments. And since these tests are for checking IDEA behaviours, I would just leave it failing, not wraping another `assertFailWith` with message check.